clang/test/SemaSYCL/non-std-layout-param.cpp

```diff
@@ -20,9 +20,61 @@ __attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
 
 
 void test() {
-  // expected-error@+1 {{kernel parameter has non-standard layout class/struct type}}
   C C0;
   C0.Y=0;
-  kernel_single_task<class MyKernel>([=] { (void)C0.Y; });
+  kernel_single_task<class MyKernel>([=] {
+    // expected-error@+1 {{kernel parameter has non-standard layout class/struct type}}
+    (void)C0.Y;
+  });
+}
+
+void test_capture_explicit_ref() {
+  int p = 0;
+  double q = 0;
+  float s = 0;
+  kernel_single_task<class kernel_capture_single_ref>([
+    // expected-error@+1 {{kernel parameter has non-standard layout class/struct type}}
+      &p,
+      q,
+    // expected-error@+1 {{kernel parameter has non-standard layout class/struct type}}
+      &s] {
+    (void) q;
+    (void) p;
+    (void) s;
+  });
 }
 
+void test_capture_implicit_refs() {
+  int p = 0;
+  double q = 0;
+  kernel_single_task<class kernel_capture_refs>([&] {
```

Comment:
Actually this case is prohibited by the 1.2.1. SYCL spec:
> The kernel function for the lambda function is the lambda function itself. The kernel lambda must use copy for
all of its captures (i.e. [=]).

Shouldn't we mention it in a diagnostic?

I also wonder which problem motivated us to made this change.

Reply:
Probably this is required to produce some values to a USM pointer or other extensions.